main.go

```diff
@@ -0,0 +1,501 @@
+		})
+	}
+}
+
+func main() {
+	var ft = flag.String("t", "", "(Required) Specifies the target structure.")
+	var fo = flag.String("o", "",
+		`This specifies the output path for the generated toml document.
+It is recommended that the location is in the same directory or
+a subdirectory of the current package. This option is
+required.`)
+	var fp = flag.String("p", "",
+		`This specifies the target package. This is only required when
+"$GOPACKAGE" is not present in the environment. Therefore, this
+can be ignored when used as a "go generate" directive. If this
+is supplied while "$GOPACKAGE" is present, this takes priority.`)
+	flag.Parse()
+
+	ep := os.Getenv("GOPACKAGE")
+
+	// Ensures that the "target" and "output" parameters are always
+	// specified. This conditionally ensures that either the
+	// parameters or the environment specify the "package"
+	// parameter.
+	if (*ft == "" || *fo == "") || (*fp == "" && ep == "") {
+		flag.Usage()
+		os.Exit(1)
+	}
+
+	p_target := ep
+	if *fp != "" {
+		p_target = *fp
+	}
+
+	// Attempt to load information about packages from the source code
+	// located in the current working directory. This does not include
+	// information from external packages. If an external package is
+	// required to build metadata for the target struct, it should
+	// be download and available in GOROOT.
+	found, err := packages.Load(
+		&packages.Config{
+			Mode: packages.NeedName |
+				packages.NeedTypes |
+				packages.NeedTypesInfo |
+				packages.NeedSyntax |
+				packages.NeedImports |
+				packages.NeedFiles,
+			Tests: false,
+		},
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	f, err := os.Create(*fo)
+	if err != nil {
+		panic(err)
+	}
+
+	context := NewContext(f)
+	for _, p := range found {
+		// "golang.org/x/tools/go/packages" and "golang.org/x/tools/go/loader"
+		// use different type definitions to represent packages. Instead of
+		// implementing functions to handle each representation of the same
+		// data, convert the data to our own representation.
+		context.Packages[p.PkgPath] =
+			&Package{
+				Path:      p.PkgPath,
+				Fset:      p.Fset,
+				Files:     p.Syntax,
+				TypesInfo: p.TypesInfo,
+			}
+	}
+
+	if p, ok := context.Packages[p_target]; ok {
+		for _, d := range p.TypesInfo.Defs {
+			if d == nil {
+				continue
+			}
+
+			// `p.TypesInfo.Defs` maps names to definitions. This can include
+			// both variables and type-names. `types.TypeName` represents the
+			// type-name for either "defined" or "aliased" types. This
+			// ensures variables are not considered.
+			//
+			// Do not use aliases! While is is possible to locate the type-
+			// definition for a type-alias (and then find the corresponding
+			// `ast.Node`), this functionality is not implemented. Just use
+			// the original struct name.
+			if _, ok := d.(*types.TypeName); !ok {
+				continue
+			}
+
+			// ensure that the `d` is struct type definition.
+			if _, ok := d.Type().Underlying().(*types.Struct); !ok {
```

Comment:
`d.Type()` is never `nil` here?

Reply:
I looked into the package and `TypeName` is a wrapper around an anonymous field. That field contains a `Type` field (instance, not pointer to instance). So, I think it should always be `non-nil` here.